src/library_stack_trace.js

```diff
@@ -12,7 +12,8 @@ var LibraryStackTrace = {
 #if DEMANGLE_SUPPORT
     // If demangle has failed before, stop demangling any further function names
     // This avoids an infinite recursion with malloc()->abort()->stackTrace()->demangle()->malloc()->...
-    if (demangle.failed) return func;
+    demangle.recursionGuard = (demangle.recursionGuard|0)+1;
+    if (demangle.recursionGuard > 1) return func;
     var __cxa_demangle_func = Module['___cxa_demangle'] || Module['__cxa_demangle'];
```

Comment:
the counting seems a little confusing. how about
```
if (demangle.recursionGuard) return;
..
try {
  demangle.recursionGuard = true;
  ...
} finally {
  demangle.recursionGuard = false;
}
```

Reply:
It needs to be like that since we need to latch-disable demangle from working when it recurses. That is, 0=not inside demangle, 1=inside demangle, 2=demangle is now disabled.

Reply:
Why not disable it on 1?

Reply:
Well, how do you disable it? You need to somehow store information to mark down that demangle is now disabled - setting the recursion guard to value 2 is that information.

Reply:
In the code sample I wrote, doesn't the line `if (demangle.recursionGuard) return;` do that? It ensures we never recurse into this function; if we enter it when it's already on the call stack from earlier then we immediately return and don't try to demangle anything.

Reply:
Oh, right. Yes, but that still has an issue that the calling demangleAll function will continue iterating over the whole callstack, and will call demangle for each function. So even though there is no infinite recursion, there can be hundreds of repeated demangle->malloc->abort calls if the stack is deep (like in the test). So on first failure, we should ensure demangle no longer tries to repeat allocation and fail again.

Reply:
May be worth putting a recursion guard in `demangleAll` then, or alternatively we could have `demangle.error` which is cleared in `demangleAll`?

Reply:
An error In demangling should not prevent demangling, only recursing to itself should prevent demangling. I don’t think it makes sense to split this check across two functions.

If you’d like, please propose a simpler patch for the failing test? I cannot come up with a better way than this. 😕

Reply:
An error In demangling should not prevent demangling, only recursing to itself should prevent demangling. I don’t think it makes sense to split this check across two functions.

If you’d like, please propose a simpler patch for the failing test? I cannot come up with a better way than this. 😕

Reply:
Doing better than this is not easy, I see what you mean.